Writing PTR records at the exact reverse-DNS key (no random suffix) is the right fix for the read/churn problem, but it breaks the **delete** path in default (non-strict) mode.

### Mechanism

PTR keys now terminate in the bare octet:

```
2.0.26.172.in-addr.arpa  → /skydns/arpa/in-addr/172/26/0/2
20.0.26.172.in-addr.arpa → /skydns/arpa/in-addr/172/26/0/20
```

`/skydns/arpa/in-addr/172/26/0/2` is a **byte-prefix** of the key for `.20` (and `.21`…`.29`, `.200`…`.255`).

The non-strict delete uses `WithPrefix` (`provider/coredns/coredns.go:218`):

```go
_, err := c.client.Delete(ctx, key, etcdcv3.WithPrefix())
```

Per etcd, `Delete(key, WithPrefix())` deletes the whole range `[key, getPrefix(key))` — i.e. every key with `key` as a byte-prefix:

- `WithPrefix()` sets `op.end = getPrefix(op.key)` — [op.go#L397](https://github.com/etcd-io/etcd/blob/v3.6.12/client/v3/op.go#L397)
- `getPrefix()` increments the last byte to form the range end — [op.go#L379](https://github.com/etcd-io/etcd/blob/v3.6.12/client/v3/op.go#L379)
- Doc: *"WithPrefix enables 'Get', 'Delete', or 'Watch' requests to operate on the keys with matching prefix"* ([etcd v3.6.12](https://github.com/etcd-io/etcd/blob/v3.6.12/client/v3/op.go#L394-L396), the version this repo pins)

So deleting `.2` computes range `["/.../0/2", "/.../0/3")`, which contains `/.../0/20`. **Deleting the PTR for `…0.2` silently wipes `…0.20`–`…0.29`, `…0.200`–`…0.255`.**

Pre-PTR this was safe because every key ended in a unique `%08x` suffix that never prefixed a sibling. The bare-octet leaf introduced here breaks that invariant.

### Why the added tests don't catch it

`fakeETCDClient`/`recordingETCDClient` implement `DeleteService` as exact-key `delete(c.services, key)`, which diverges from the production `WithPrefix` range delete. A fake that mirrors etcd reproduces the bug:

```go
// models real non-strict etcdClient.DeleteService: Delete(key, WithPrefix())
// = range delete [key, getPrefix(key)) == all keys byte-prefixed by `key`.
type prefixDeletingETCDClient struct{ services map[string]Service }

func (c *prefixDeletingETCDClient) GetServices(_ context.Context, prefix string) ([]*Service, error) {
	var result []*Service
	for key, value := range c.services {
		if strings.HasPrefix(key, prefix) {
			v := value
			v.Key = key
			result = append(result, &v)
		}
	}
	return result, nil
}
func (c *prefixDeletingETCDClient) SaveService(_ context.Context, s *Service) error {
	c.services[s.Key] = *s
	return nil
}
func (c *prefixDeletingETCDClient) DeleteService(_ context.Context, key string) error {
	for k := range c.services {
		if strings.HasPrefix(k, key) { // etcdcv3.WithPrefix semantics
			delete(c.services, k)
		}
	}
	return nil
}

func TestPTRDeleteOverDeletesNeighbors(t *testing.T) {
	client := &prefixDeletingETCDClient{services: map[string]Service{}}
	coredns := coreDNSProvider{client: client, coreDNSPrefix: defaultCoreDNSPrefix}

	require.NoError(t, coredns.ApplyChanges(t.Context(), &plan.Changes{
		Create: []*endpoint.Endpoint{
			endpoint.NewEndpoint("2.0.26.172.in-addr.arpa", endpoint.RecordTypePTR, "host-2"),
			endpoint.NewEndpoint("20.0.26.172.in-addr.arpa", endpoint.RecordTypePTR, "host-20"),
		},
	}))

	key2 := "/skydns/arpa/in-addr/172/26/0/2"
	key20 := "/skydns/arpa/in-addr/172/26/0/20"
	require.Contains(t, client.services, key2)
	require.Contains(t, client.services, key20)

	// delete ONLY .2
	require.NoError(t, coredns.ApplyChanges(t.Context(), &plan.Changes{
		Delete: []*endpoint.Endpoint{
			endpoint.NewEndpoint("2.0.26.172.in-addr.arpa", endpoint.RecordTypePTR, "host-2"),
		},
	}))

	assert.NotContains(t, client.services, key2) // intended
	assert.Contains(t, client.services, key20,   // FAILS today: collateral delete
		"deleting .2 must NOT delete neighbour .20 — etcd WithPrefix over-delete")
}
```

Result on the current branch — after deleting `.2` the store is **empty**, `.20` gone too:

```
--- FAIL: TestPTRDeleteOverDeletesNeighbors (0.00s)
    coredns_test.go: Error: map[string]coredns.Service{} does not contain "/skydns/arpa/in-addr/172/26/0/20"
    Messages: deleting .2 must NOT delete neighbour .20 — etcd WithPrefix over-delete
```

@dongjiang1989 Maybe you can delete PTR leaf keys with an exact match (no WithPrefix), since they have no children ? (and re-work test to use a more accurate fake on ETCDClient).

--- a/provider/coredns/coredns.go
+++ b/provider/coredns/coredns.go
@@ -53,7 +53,7 @@ const (
 var (
 	// avoids allocating a new slice on every call
 	// prevents unwanted deletion of etcd keys based on labels
-	skipLabels = []string{originalTextLabel, "prefix", "resource", endpoint.OwnerLabelKey}
+	skipLabels = []string{originalTextLabel, randomPrefixLabel, "resource", endpoint.OwnerLabelKey}
 )
 
 // coreDNSClient is an interface to work with CoreDNS service records in etcd

--- a/provider/coredns/coredns.go
+++ b/provider/coredns/coredns.go
@@ -47,19 +47,20 @@ const (
 
 	randomPrefixLabel     = "prefix"
 	providerSpecificGroup = "coredns/group"
+	originalTextLabel     = "originalText"
 )
 
 var (
 	// avoids allocating a new slice on every call
 	// prevents unwanted deletion of etcd keys based on labels
-	skipLabels = []string{"originalText", "prefix", "resource", endpoint.OwnerLabelKey}
+	skipLabels = []string{originalTextLabel, "prefix", "resource", endpoint.OwnerLabelKey}
 )
 
 // coreDNSClient is an interface to work with CoreDNS service records in etcd
 type coreDNSClient interface {
 	GetServices(ctx context.Context, prefix string) ([]*Service, error)
 	SaveService(ctx context.Context, value *Service) error
-	DeleteService(ctx context.Context, key string) error
+	DeleteService(ctx context.Context, key string, exact bool) error
 }
 
 type coreDNSProvider struct {
@@ -188,13 +189,19 @@ func (c etcdClient) SaveService(ctx context.Context, service *Service) error {
 	return nil
 }
 
-// DeleteService deletes service record from etcd
-func (c etcdClient) DeleteService(ctx context.Context, key string) error {
+// DeleteService deletes service record from etcd.
+// If exact is true, only the single key is deleted. Otherwise all keys sharing
+// the key as a byte-prefix are removed (etcd WithPrefix semantics).
+func (c etcdClient) DeleteService(ctx context.Context, key string, exact bool) error {
 	ctx, cancel := context.WithTimeout(ctx, etcdTimeout)
 	defer cancel()
 
 	if c.strictlyOwned {
-		rs, err := c.client.Get(ctx, key, etcdcv3.WithPrefix())
+		var getOpts []etcdcv3.OpOption
+		if !exact {
+			getOpts = append(getOpts, etcdcv3.WithPrefix())
+		}
+		rs, err := c.client.Get(ctx, key, getOpts...)
 		if err != nil {
 			return err
 		}
@@ -213,10 +220,13 @@ func (c etcdClient) DeleteService(ctx context.Context, key string) error {
 			}
 		}
 		return err
-	} else {
-		_, err := c.client.Delete(ctx, key, etcdcv3.WithPrefix())
+	}
+	if exact {
+		_, err := c.client.Delete(ctx, key)
 		return err
 	}
+	_, err := c.client.Delete(ctx, key, etcdcv3.WithPrefix())
+	return err
 }
 
 func (c etcdClient) unmarshalService(n *mvccpb.KeyValue) (*Service, error) {
@@ -325,9 +335,13 @@ func (p coreDNSProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, err
 				ep.Targets = append(ep.Targets, service.Host)
 				log.Debugf("Extending ep (%s) with new service host (%s)", ep, service.Host)
 			} else {
+				recordType := guessRecordType(service.Host)
+				if isPTRDomain(dnsName) {
+					recordType = endpoint.RecordTypePTR
+				}
 				ep = endpoint.NewEndpointWithTTL(
 					dnsName,
-					guessRecordType(service.Host),
+					recordType,
 					endpoint.TTL(service.TTL),
 					service.Host,
 				)
@@ -339,7 +353,7 @@ func (p coreDNSProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, err
 			if p.strictlyOwned {
 				ep.Labels[endpoint.OwnerLabelKey] = service.Owner
 			}
-			ep.Labels["originalText"] = service.Text
+			ep.Labels[originalTextLabel] = service.Text
 			ep.Labels[randomPrefixLabel] = prefix
 			ep.Labels[service.Host] = prefix
 			result = append(result, ep)
@@ -421,6 +435,19 @@ func (p coreDNSProvider) createServicesForEndpoint(ctx context.Context, dnsName 
 	var services []*Service
 
 	for _, target := range ep.Targets {
+		if ep.RecordType == endpoint.RecordTypePTR {
+			// CoreDNS etcd plugin expects PTR records at the exact reverse-DNS
+			// path (e.g., /skydns/arpa/in-addr/172/26/0/2) without any unique suffix.
+			services = append(services, &Service{
+				Host:        target,
+				Text:        ep.Labels[originalTextLabel],
+				Key:         p.etcdKeyFor(dnsName),
+				TargetStrip: 0,
+				TTL:         uint32(ep.RecordTTL),
+			})
+			continue
+		}
+
 		prefix := ep.Labels[target]
 		if prefix == "" {
 			prefix = fmt.Sprintf("%08x", rand.Int31())
@@ -432,7 +459,7 @@ func (p coreDNSProvider) createServicesForEndpoint(ctx context.Context, dnsName 
 		}
 		service := Service{
 			Host:        target,
-			Text:        ep.Labels["originalText"],
+			Text:        ep.Labels[originalTextLabel],
 			Key:         p.etcdKeyFor(prefix + "." + dnsName),
 			TargetStrip: strings.Count(prefix, ".") + 1,
 			TTL:         uint32(ep.RecordTTL),
@@ -442,14 +469,21 @@ func (p coreDNSProvider) createServicesForEndpoint(ctx context.Context, dnsName 
 		ep.Labels[target] = prefix
 	}
 
-	// Clean outdated labels
+	// Clean outdated labels.
+	// For PTR records, all targets share the same etcd key (the exact
+	// reverse-DNS path). Stale labels are superseded by the in-place overwrite
+	// from the new target — issuing a delete would either be a no-op (save
+	// restores the key) or, worse, delete a sibling key that shares the same
+	// byte-prefix. So skip stale-label cleanup entirely for PTR records.
+	if ep.RecordType == endpoint.RecordTypePTR {
+		return services, nil
+	}
 	for label, labelPrefix := range ep.Labels {
 		if slices.Contains(skipLabels, label) {
 			continue
 		}
 		if !slices.Contains(ep.Targets, label) {
-			err := p.deleteByDnsName(ctx, labelPrefix+"."+dnsName)
-			if err != nil {
+			if err := p.deleteByDnsName(ctx, labelPrefix+"."+dnsName, false); err != nil {
 				return nil, err
 			}
 		}
@@ -459,19 +493,31 @@ func (p coreDNSProvider) createServicesForEndpoint(ctx context.Context, dnsName 
 
 // updateTXTRecords updates the TXT records in the provided services slice based on the given group of endpoints.
 func (p coreDNSProvider) updateTXTRecords(dnsName string, group []*endpoint.Endpoint, services []*Service) []*Service {
+	isPTR := isPTRDomain(dnsName)
 	index := 0
 	for _, ep := range group {
 		if ep.RecordType != endpoint.RecordTypeTXT {
 			continue
 		}
 		if index >= len(services) {
-			prefix := ep.Labels[randomPrefixLabel]
-			if prefix == "" {
-				prefix = fmt.Sprintf("%08x", rand.Int31())
+			var key string
+			var targetStrip int
+			if isPTR {
+				// PTR reverse zones must not have a random prefix — CoreDNS
+				// expects the key at the exact reverse-DNS path.
+				key = p.etcdKeyFor(dnsName)
+				targetStrip = 0
+			} else {
+				prefix := ep.Labels[randomPrefixLabel]
+				if prefix == "" {
+					prefix = fmt.Sprintf("%08x", rand.Int31())
+				}
+				key = p.etcdKeyFor(prefix + "." + dnsName)
+				targetStrip = strings.Count(prefix, ".") + 1
 			}
 			services = append(services, &Service{
-				Key:         p.etcdKeyFor(prefix + "." + dnsName),
-				TargetStrip: strings.Count(prefix, ".") + 1,
+				Key:         key,
+				TargetStrip: targetStrip,
 				TTL:         uint32(ep.RecordTTL),
 			})
 		}
@@ -491,7 +537,8 @@ func (p coreDNSProvider) deleteEndpoints(ctx context.Context, endpoints []*endpo
 		if ep.Labels[randomPrefixLabel] != "" {
 			dnsName = ep.Labels[randomPrefixLabel] + "." + dnsName
 		}
-		err := p.deleteByDnsName(ctx, dnsName)
+		exact := ep.RecordType == endpoint.RecordTypePTR || ep.RecordType == endpoint.RecordTypeTXT
+		err := p.deleteByDnsName(ctx, dnsName, exact)
 		if err != nil {
 			return err
 		}
@@ -499,13 +546,13 @@ func (p coreDNSProvider) deleteEndpoints(ctx context.Context, endpoints []*endpo
 	return nil
 }
 
-func (p coreDNSProvider) deleteByDnsName(ctx context.Context, dnsName string) error {
+func (p coreDNSProvider) deleteByDnsName(ctx context.Context, dnsName string, exact bool) error {
 	key := p.etcdKeyFor(dnsName)
-	log.Infof("Delete key %s", key)
+	log.Infof("Delete key %s (exact=%v)", key, exact)
 	if p.dryRun {
 		return nil
 	}
-	if err := p.client.DeleteService(ctx, key); err != nil {
+	if err := p.client.DeleteService(ctx, key, exact); err != nil {
 		return err
 	}
 	return nil
@@ -515,6 +562,10 @@ func (p coreDNSProvider) etcdKeyFor(dnsName string) string {
 	domains := strings.Split(dnsName, ".")
 	reverse(domains)
 	return p.coreDNSPrefix + strings.Join(domains, "/")
+}
+
+func isPTRDomain(dnsName string) bool {
+	return strings.HasSuffix(dnsName, ".in-addr.arpa") || strings.HasSuffix(dnsName, ".ip6.arpa")
 }
 
 func guessRecordType(target string) string {

--- a/provider/coredns/coredns.go
+++ b/provider/coredns/coredns.go
@@ -436,6 +436,10 @@ func (p coreDNSProvider) createServicesForEndpoint(ctx context.Context, dnsName 
 
 	for _, target := range ep.Targets {
 		if ep.RecordType == endpoint.RecordTypePTR {
+			if len(ep.Targets) > 1 {
+				log.Warnf("PTR %s has %d targets; CoreDNS etcd backend stores one target per reverse-DNS key, using %q and ignoring %v",
+					dnsName, len(ep.Targets), ep.Targets[0], ep.Targets[1:])
+			}
 			// CoreDNS etcd plugin expects PTR records at the exact reverse-DNS
 			// path (e.g., /skydns/arpa/in-addr/172/26/0/2) without any unique suffix.
 			services = append(services, &Service{

--- a/provider/coredns/coredns_test.go
+++ b/provider/coredns/coredns_test.go
@@ -41,10 +41,11 @@ import (
 const defaultCoreDNSPrefix = "/skydns/"
 
 type fakeETCDClient struct {
-	services map[string]Service
+	services    map[string]Service
+	deletedKeys []string
 }
 
-func (c fakeETCDClient) GetServices(_ context.Context, prefix string) ([]*Service, error) {
+func (c *fakeETCDClient) GetServices(_ context.Context, prefix string) ([]*Service, error) {
 	var result []*Service
 	for key, value := range c.services {
 		if strings.HasPrefix(key, prefix) {
@@ -56,13 +57,26 @@ func (c fakeETCDClient) GetServices(_ context.Context, prefix string) ([]*Servic
 	return result, nil
 }
 
-func (c fakeETCDClient) SaveService(_ context.Context, service *Service) error {
+func (c *fakeETCDClient) SaveService(_ context.Context, service *Service) error {
 	c.services[service.Key] = *service
 	return nil
 }
 
-func (c fakeETCDClient) DeleteService(_ context.Context, key string) error {
-	delete(c.services, key)
+// DeleteService mirrors real etcd DeleteService semantics:
+// exact=false uses WithPrefix (range delete of all keys byte-prefixed by key),
+// exact=true deletes only the single key.
+// Also records all keys passed to DeleteService for test assertions.
+func (c *fakeETCDClient) DeleteService(_ context.Context, key string, exact bool) error {
+	c.deletedKeys = append(c.deletedKeys, key)
+	if exact {
+		delete(c.services, key)
+		return nil
+	}
+	for k := range c.services {
+		if strings.HasPrefix(k, key) {
+			delete(c.services, k)
+		}
+	}
 	return nil
 }
 
@@ -171,10 +185,9 @@ func TestAServiceTranslation(t *testing.T) {
 	expectedDNSName := "example.com"
 	expectedRecordType := endpoint.RecordTypeA
 
-	client := fakeETCDClient{
-		map[string]Service{
-			"/skydns/com/example": {Host: expectedTarget},
-		},
+	client := &fakeETCDClient{services: map[string]Service{
+		"/skydns/com/example": {Host: expectedTarget},
+	},
 	}
 	provider := coreDNSProvider{
 		client:        client,
@@ -218,9 +231,7 @@ func TestGuessRecordType(t *testing.T) {
 // It asserts both record types survive an apply + read round-trip with the correct
 // type (A stays A, AAAA stays AAAA), and that a second reconcile is a no-op.
 func TestCoreDNSRoundTrip(t *testing.T) {
-	client := fakeETCDClient{
-		map[string]Service{},
-	}
+	client := &fakeETCDClient{services: map[string]Service{}}
 	coredns := coreDNSProvider{
 		client:        client,
 		coreDNSPrefix: defaultCoreDNSPrefix,
@@ -270,10 +281,9 @@ func TestCNAMEServiceTranslation(t *testing.T) {
 	expectedDNSName := "example.com"
 	expectedRecordType := endpoint.RecordTypeCNAME
 
-	client := fakeETCDClient{
-		map[string]Service{
-			"/skydns/com/example": {Host: expectedTarget},
-		},
+	client := &fakeETCDClient{services: map[string]Service{
+		"/skydns/com/example": {Host: expectedTarget},
+	},
 	}
 	provider := coreDNSProvider{
 		client:        client,
@@ -300,10 +310,9 @@ func TestTXTServiceTranslation(t *testing.T) {
 	expectedDNSName := "example.com"
 	expectedRecordType := endpoint.RecordTypeTXT
 
-	client := fakeETCDClient{
-		map[string]Service{
-			"/skydns/com/example": {Text: expectedTarget},
-		},
+	client := &fakeETCDClient{services: map[string]Service{
+		"/skydns/com/example": {Text: expectedTarget},
+	},
 	}
 	provider := coreDNSProvider{
 		client:        client,
@@ -332,10 +341,9 @@ func TestAWithTXTServiceTranslation(t *testing.T) {
 	}
 	expectedDNSName := "example.com"
 
-	client := fakeETCDClient{
-		map[string]Service{
-			"/skydns/com/example": {Host: "1.2.3.4", Text: "string"},
-		},
+	client := &fakeETCDClient{services: map[string]Service{
+		"/skydns/com/example": {Host: "1.2.3.4", Text: "string"},
+	},
 	}
 	provider := coreDNSProvider{
 		client:        client,
@@ -372,10 +380,9 @@ func TestCNAMEWithTXTServiceTranslation(t *testing.T) {
 	}
 	expectedDNSName := "example.com"
 
-	client := fakeETCDClient{
-		map[string]Service{
-			"/skydns/com/example": {Host: "example.net", Text: "string"},
-		},
+	client := &fakeETCDClient{services: map[string]Service{
+		"/skydns/com/example": {Host: "example.net", Text: "string"},
+	},
 	}
 	provider := coreDNSProvider{
 		client:        client,
@@ -405,10 +412,62 @@ func TestCNAMEWithTXTServiceTranslation(t *testing.T) {
 	}
 }
 
-func TestCoreDNSApplyChanges(t *testing.T) {
-	client := fakeETCDClient{
-		map[string]Service{},
+// Deleting a TXT ownership marker must NOT delete the sibling A record that
+// shares the same dnsName. TXT keys (e.g. /skydns/local/domain1) are byte-
+// prefixes of A keys (/skydns/local/domain1/<random>); WithPrefix delete on
+// the TXT key would range-delete the A key too.
+func TestDeleteTXTDoesNotOverDeleteARecord(t *testing.T) {
+	client := &fakeETCDClient{services: map[string]Service{}}
+	coredns := coreDNSProvider{client: client, coreDNSPrefix: defaultCoreDNSPrefix}
+
+	// Create A + TXT for domain1, and A for domain2
+	err := coredns.ApplyChanges(t.Context(), &plan.Changes{
+		Create: []*endpoint.Endpoint{
+			endpoint.NewEndpoint("domain1.local", endpoint.RecordTypeA, "5.5.5.5"),
+			endpoint.NewEndpoint("domain1.local", endpoint.RecordTypeTXT, "owner-marker"),
+			endpoint.NewEndpoint("domain2.local", endpoint.RecordTypeA, "6.6.6.6"),
+		},
+	})
+	require.NoError(t, err)
+
+	// Verify A record exists for domain1 before delete
+	foundBefore := false
+	for _, svc := range client.services {
+		if svc.Host == "5.5.5.5" {
+			foundBefore = true
+		}
 	}
+	require.True(t, foundBefore, "A record must exist before TXT delete")
+
+	// Delete ONLY the TXT for domain1 (not the A record)
+	err = coredns.ApplyChanges(t.Context(), &plan.Changes{
+		Delete: []*endpoint.Endpoint{
+			endpoint.NewEndpoint("domain1.local", endpoint.RecordTypeTXT, "owner-marker"),
+		},
+	})
+	require.NoError(t, err)
+
+	// domain1's A record must survive
+	foundA := false
+	for _, svc := range client.services {
+		if svc.Host == "5.5.5.5" {
+			foundA = true
+		}
+	}
+	assert.True(t, foundA, "TXT delete must NOT delete sibling A record (WithPrefix over-delete)")
+
+	// domain2's A record must also survive
+	foundD2 := false
+	for _, svc := range client.services {
+		if svc.Host == "6.6.6.6" {
+			foundD2 = true
+		}
+	}
+	assert.True(t, foundD2, "domain2 A record must be unaffected")
+}
+
+func TestCoreDNSApplyChanges(t *testing.T) {
+	client := &fakeETCDClient{services: map[string]Service{}}
 	coredns := coreDNSProvider{
 		client:        client,
 		coreDNSPrefix: defaultCoreDNSPrefix,
@@ -489,9 +548,7 @@ func TestCoreDNSApplyChanges(t *testing.T) {
 }
 
 func TestCoreDNSApplyChanges_DomainDoNotMatch(t *testing.T) {
-	client := fakeETCDClient{
-		map[string]Service{},
-	}
+	client := &fakeETCDClient{services: map[string]Service{}}
 	coredns := coreDNSProvider{
 		client:        client,
 		coreDNSPrefix: defaultCoreDNSPrefix,
@@ -823,7 +880,7 @@ func TestDeleteService(t *testing.T) {
 				},
 			}
 
-			err := c.DeleteService(t.Context(), tt.key)
+			err := c.DeleteService(t.Context(), tt.key, false)
 
 			if tt.wantErr {
 				require.Error(t, err)
@@ -962,7 +1019,7 @@ func TestDeleteServiceWithStrictlyOwned(t *testing.T) {
 				strictlyOwned: true,
 			}
 
-			err := c.DeleteService(t.Context(), tt.key)
+			err := c.DeleteService(t.Context(), tt.key, false)
 
 			require.NoError(t, err)
 			mockKV.AssertExpectations(t)
@@ -1370,9 +1427,7 @@ func TestCoreDNSProvider_updateTXTRecords_ClearsExtraText(t *testing.T) {
 }
 
 func TestApplyChangesAWithGroupServiceTranslation(t *testing.T) {
-	client := fakeETCDClient{
-		map[string]Service{},
-	}
+	client := &fakeETCDClient{services: map[string]Service{}}
 	coredns := coreDNSProvider{
 		client:        client,
 		coreDNSPrefix: defaultCoreDNSPrefix,
@@ -1397,14 +1452,13 @@ func TestApplyChangesAWithGroupServiceTranslation(t *testing.T) {
 
 // Prevents unwanted deletion of etcd keys https://github.com/kubernetes-sigs/external-dns/commit/83ea480c57fcc6beeaf8a6a3e8446cb87e07415d
 func TestApplyChangesPreventCleanupForKnownLabels(t *testing.T) {
-	client := fakeETCDClient{
-		map[string]Service{
-			"/skydns/local/domain1/test":         {Host: "10.0.0.0"},
-			"/skydns/local/domain1/originalText": {Host: "10.0.0.0"},
-			"/skydns/local/domain1/prefix":       {Host: "10.0.0.0"},
-			"/skydns/local/domain1/resource":     {Host: "10.0.0.0"},
-			"/skydns/local/domain1/owner":        {Host: "10.0.0.0"},
-		},
+	client := &fakeETCDClient{services: map[string]Service{
+		"/skydns/local/domain1/test":         {Host: "10.0.0.0"},
+		"/skydns/local/domain1/originalText": {Host: "10.0.0.0"},
+		"/skydns/local/domain1/prefix":       {Host: "10.0.0.0"},
+		"/skydns/local/domain1/resource":     {Host: "10.0.0.0"},
+		"/skydns/local/domain1/owner":        {Host: "10.0.0.0"},
+	},
 	}
 	coredns := coreDNSProvider{
 		client:        client,
@@ -1456,10 +1510,9 @@ func TestApplyChangesPreventCleanupForKnownLabels(t *testing.T) {
 }
 
 func TestRecordsAWithGroupServiceTranslation(t *testing.T) {
-	client := fakeETCDClient{
-		map[string]Service{
-			"/skydns/local/domain1": {Host: "5.5.5.5", Group: "test1"},
-		},
+	client := &fakeETCDClient{services: map[string]Service{
+		"/skydns/local/domain1": {Host: "5.5.5.5", Group: "test1"},
+	},
 	}
 	coredns := coreDNSProvider{
 		client:        client,
@@ -1475,11 +1528,10 @@ func TestRecordsAWithGroupServiceTranslation(t *testing.T) {
 }
 
 func TestRecordsIncludeLabelOwnerWithStrictlyOwned(t *testing.T) {
-	client := fakeETCDClient{
-		map[string]Service{
-			"/skydns/local/domain1": {Host: "5.5.5.5", Group: "test1", Owner: "owner"},
-			"/skydns/com/example":   {Text: "bla", Owner: "owner"},
-		},
+	client := &fakeETCDClient{services: map[string]Service{
+		"/skydns/local/domain1": {Host: "5.5.5.5", Group: "test1", Owner: "owner"},
+		"/skydns/com/example":   {Text: "bla", Owner: "owner"},
+	},
 	}
 	coredns := coreDNSProvider{
 		client:        client,
@@ -1494,11 +1546,10 @@ func TestRecordsIncludeLabelOwnerWithStrictlyOwned(t *testing.T) {
 }
 
 func TestRecordsIncludeOwnerASLabelWithoutStrictlyOwned(t *testing.T) {
-	client := fakeETCDClient{
-		map[string]Service{
-			"/skydns/local/domain1": {Host: "5.5.5.5", Group: "test1", Owner: "owner"},
-			"/skydns/com/example":   {Text: "bla", Owner: "owner"},
-		},
+	client := &fakeETCDClient{services: map[string]Service{
+		"/skydns/local/domain1": {Host: "5.5.5.5", Group: "test1", Owner: "owner"},
+		"/skydns/com/example":   {Text: "bla", Owner: "owner"},
+	},
 	}
 	coredns := coreDNSProvider{
 		client:        client,
@@ -1510,4 +1561,235 @@ func TestRecordsIncludeOwnerASLabelWithoutStrictlyOwned(t *testing.T) {
 	for _, ep := range endpoints {
 		assert.Empty(t, ep.Labels[endpoint.OwnerLabelKey])
 	}
+}
+
+func TestPTRRecords(t *testing.T) {
+	t.Run("ReadIPv4PTR", func(t *testing.T) {
+		expectedTarget := "coredns-etcd-worker"
+		expectedDNSName := "2.0.26.172.in-addr.arpa"
+
+		client := &fakeETCDClient{services: map[string]Service{
+			"/skydns/arpa/in-addr/172/26/0/2": {Host: expectedTarget},
+		},
+		}
+		provider := coreDNSProvider{
+			client:        client,
+			coreDNSPrefix: defaultCoreDNSPrefix,
+		}
+		endpoints, err := provider.Records(t.Context())
+		require.NoError(t, err)
+		require.Len(t, endpoints, 1)
+		assert.Equal(t, expectedDNSName, endpoints[0].DNSName)
+		assert.Equal(t, expectedTarget, endpoints[0].Targets[0])
+		assert.Equal(t, endpoint.RecordTypePTR, endpoints[0].RecordType)
+	})
+
+	t.Run("ReadIPv6PTR", func(t *testing.T) {
+		expectedTarget := "coredns-etcd-worker"
+		// IPv6 2001:db8::1 → 32 nibbles reversed
+		expectedDNSName := "1.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.8.b.d.0.1.0.0.2.ip6.arpa"
+
+		client := &fakeETCDClient{services: map[string]Service{
+			"/skydns/arpa/ip6/2/0/0/1/0/d/b/8/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/1": {Host: expectedTarget},
+		},
+		}
+		provider := coreDNSProvider{
+			client:        client,
+			coreDNSPrefix: defaultCoreDNSPrefix,
+		}
+		endpoints, err := provider.Records(t.Context())
+		require.NoError(t, err)
+		require.Len(t, endpoints, 1)
+		assert.Equal(t, expectedDNSName, endpoints[0].DNSName)
+		assert.Equal(t, expectedTarget, endpoints[0].Targets[0])
+		assert.Equal(t, endpoint.RecordTypePTR, endpoints[0].RecordType)
+	})
+
+	t.Run("CreateAndDeletePTR", func(t *testing.T) {
+		client := &fakeETCDClient{services: map[string]Service{}}
+		coredns := coreDNSProvider{
+			client:        client,
+			coreDNSPrefix: defaultCoreDNSPrefix,
+		}
+
+		// Create PTR records
+		changes := &plan.Changes{
+			Create: []*endpoint.Endpoint{
+				endpoint.NewEndpoint("2.0.26.172.in-addr.arpa", endpoint.RecordTypePTR, "coredns-etcd-worker"),
+				endpoint.NewEndpoint("3.0.26.172.in-addr.arpa", endpoint.RecordTypePTR, "coredns-etcd-control-plane"),
+			},
+		}
+		err := coredns.ApplyChanges(t.Context(), changes)
+		require.NoError(t, err)
+
+		// Verify keys are created WITHOUT random prefix
+		expectedServices := map[string][]*Service{
+			"/skydns/arpa/in-addr/172/26/0/2": {{Host: "coredns-etcd-worker", TargetStrip: 0}},
+			"/skydns/arpa/in-addr/172/26/0/3": {{Host: "coredns-etcd-control-plane", TargetStrip: 0}},
+		}
+		validateServices(client.services, expectedServices, t, 1)
+
+		// Delete PTR records
+		deleteChanges := &plan.Changes{
+			Delete: []*endpoint.Endpoint{
+				endpoint.NewEndpoint("2.0.26.172.in-addr.arpa", endpoint.RecordTypePTR, "coredns-etcd-worker"),
+			},
+		}
+		err = applyServiceChanges(coredns, deleteChanges)
+		require.NoError(t, err)
+
+		// Only the second record should remain
+		expectedAfterDelete := map[string][]*Service{
+			"/skydns/arpa/in-addr/172/26/0/3": {{Host: "coredns-etcd-control-plane", TargetStrip: 0}},
+		}
+		validateServices(client.services, expectedAfterDelete, t, 2)
+	})
+
+	t.Run("UpdatePTRTargetOverwritesInPlace", func(t *testing.T) {
+		// Verifies that updating a PTR target overwrites the same key (no
+		// separate delete needed — PTR records share a single etcd key, so
+		// stale-label cleanup is skipped).
+		client := &fakeETCDClient{services: map[string]Service{}}
+		coredns := coreDNSProvider{
+			client:        client,
+			coreDNSPrefix: defaultCoreDNSPrefix,
+		}
+
+		// Step 1: Create a PTR record pointing to "old-host"
+		createChanges := &plan.Changes{
+			Create: []*endpoint.Endpoint{
+				endpoint.NewEndpoint("2.0.26.172.in-addr.arpa", endpoint.RecordTypePTR, "old-host"),
+			},
+		}
+		require.NoError(t, coredns.ApplyChanges(t.Context(), createChanges))
+
+		key := "/skydns/arpa/in-addr/172/26/0/2"
+		require.Equal(t, "old-host", client.services[key].Host)
+
+		// Step 2: Update target to "new-host"
+		records, err := coredns.Records(t.Context())
+		require.NoError(t, err)
+		oldEP := records[0]
+
+		newEP := endpoint.NewEndpoint("2.0.26.172.in-addr.arpa", endpoint.RecordTypePTR, "new-host")
+		require.NoError(t, coredns.ApplyChanges(t.Context(), &plan.Changes{
+			UpdateNew: []*endpoint.Endpoint{newEP},
+			UpdateOld: []*endpoint.Endpoint{oldEP},
+		}))
+
+		// The key still exists with the new target (overwrite, not delete+create)
+		require.Contains(t, client.services, key)
+		assert.Equal(t, "new-host", client.services[key].Host)
+	})
+
+	t.Run("DeletePTRDoesNotOverDeleteNeighbors", func(t *testing.T) {
+		client := &fakeETCDClient{services: map[string]Service{}}
+		coredns := coreDNSProvider{client: client, coreDNSPrefix: defaultCoreDNSPrefix}
+
+		// Create PTR records for .0.2 and .0.20 (bare octet keys where .2 is
+		// a byte-prefix of .20)
+		require.NoError(t, coredns.ApplyChanges(t.Context(), &plan.Changes{
+			Create: []*endpoint.Endpoint{
+				endpoint.NewEndpoint("2.0.26.172.in-addr.arpa", endpoint.RecordTypePTR, "host-2"),
+				endpoint.NewEndpoint("20.0.26.172.in-addr.arpa", endpoint.RecordTypePTR, "host-20"),
+			},
+		}))
+
+		key2 := "/skydns/arpa/in-addr/172/26/0/2"
+		key20 := "/skydns/arpa/in-addr/172/26/0/20"
+		require.Contains(t, client.services, key2)
+		require.Contains(t, client.services, key20)
+
+		// Delete ONLY .2
+		require.NoError(t, coredns.ApplyChanges(t.Context(), &plan.Changes{
+			Delete: []*endpoint.Endpoint{
+				endpoint.NewEndpoint("2.0.26.172.in-addr.arpa", endpoint.RecordTypePTR, "host-2"),
+			},
+		}))
+
+		assert.NotContains(t, client.services, key2, "intended target must be deleted")
+		assert.Contains(t, client.services, key20,
+			"deleting .2 must NOT delete neighbour .20 — etcd WithPrefix over-delete")
+	})
+
+	t.Run("MultiTargetPTRUpdateDoesNotDeleteSurvivor", func(t *testing.T) {
+		// Scenario: PTR has targets [host-1, host-2], updated to [host-2].
+		// The stale-label cleanup must NOT delete the key that host-2 lives on.
+		client := &fakeETCDClient{services: map[string]Service{}}
+		coredns := coreDNSProvider{client: client, coreDNSPrefix: defaultCoreDNSPrefix}
+
+		// Simulate the "before" state: a PTR record at the key with host-1
+		key := "/skydns/arpa/in-addr/172/26/0/2"
+		client.services[key] = Service{
+			Host:        "host-1",
+			TargetStrip: 0,
+		}
+
+		// Read records to populate labels (simulates what the controller does)
+		records, err := coredns.Records(t.Context())
+		require.NoError(t, err)
+		require.Len(t, records, 1)
+		oldEP := records[0]
+		require.Equal(t, "host-1", oldEP.Targets[0])
+
+		// Update: change target from host-1 to host-2
+		newEP := endpoint.NewEndpoint("2.0.26.172.in-addr.arpa", endpoint.RecordTypePTR, "host-2")
+		updateChanges := &plan.Changes{
+			UpdateNew: []*endpoint.Endpoint{newEP},
+			UpdateOld: []*endpoint.Endpoint{oldEP},
+		}
+		require.NoError(t, coredns.ApplyChanges(t.Context(), updateChanges))
+
+		// host-2 should be at the key (the update should have taken effect)
+		require.Contains(t, client.services, key,
+			"PTR key must survive after updating target from host-1 to host-2")
+		assert.Equal(t, "host-2", client.services[key].Host)
+	})
+
+	t.Run("PTRStaleLabelCleanupSkippedWhenOtherTargetsRemain", func(t *testing.T) {
+		// Verifies that stale-label cleanup for PTR records does NOT delete the
+		// shared key when other targets still exist. Uses a delete-recording
+		// client to assert that deleteByDnsName is never called for PTR stale
+		// labels (PTR records are overwritten in-place, not deleted+recreated).
+		rec := &fakeETCDClient{services: map[string]Service{
+			"/skydns/arpa/in-addr/172/26/0/2": {Host: "host-1", TargetStrip: 0},
+		}}
+		coredns := coreDNSProvider{client: rec, coreDNSPrefix: defaultCoreDNSPrefix}
+
+		// Read records to get labels, then update target from host-1 → host-2
+		records, err := coredns.Records(t.Context())
+		require.NoError(t, err)
+		oldEP := records[0]
+
+		newEP := endpoint.NewEndpoint("2.0.26.172.in-addr.arpa", endpoint.RecordTypePTR, "host-2")
+		updateChanges := &plan.Changes{
+			UpdateNew: []*endpoint.Endpoint{newEP},
+			UpdateOld: []*endpoint.Endpoint{oldEP},
+		}
+		require.NoError(t, coredns.ApplyChanges(t.Context(), updateChanges))
+
+		// The delete should NOT have been called for the PTR dnsName
+		// (stale host-1 label should be cleaned up by overwrite, not by delete)
+		ptrKey := "/skydns/arpa/in-addr/172/26/0/2"
+		assert.NotContains(t, rec.deletedKeys, ptrKey,
+			"stale-label cleanup for PTR must not issue delete when new target will overwrite the same key")
+	})
+
+	t.Run("UpdateTXTRecordsForPTRDomain", func(t *testing.T) {
+		provider := coreDNSProvider{coreDNSPrefix: "/skydns/"}
+		dnsName := "3.0.18.172.in-addr.arpa"
+		group := []*endpoint.Endpoint{
+			{
+				RecordType: endpoint.RecordTypeTXT,
+				Targets:    endpoint.Targets{"txt-value"},
+				Labels:     map[string]string{randomPrefixLabel: ""},
+				RecordTTL:  60,
+			},
+		}
+		services := provider.updateTXTRecords(dnsName, group, []*Service{})
+		require.Len(t, services, 1)
+		assert.Equal(t, "/skydns/arpa/in-addr/172/18/0/3", services[0].Key)
+		assert.Equal(t, 0, services[0].TargetStrip)
+		assert.Equal(t, "txt-value", services[0].Text)
+	})
 }

